tlatools/org.lamport.tlatools/src/tlc2/tool/queue/DiskByteArrayQueue.java

```diff
@@ -738,19 +739,19 @@ public final byte[] toByteArray() {
 		 */
 		@Override
 		public final void writeString(String str) throws IOException {
-			final int length = str.length();
-			ensureCapacity(idx + length);
-			
-			final char[] c = new char[length];
-			str.getChars(0, length, c, 0);
-			
-			for (int i = 0; i < c.length; i++) {
-				this.bytes[idx++] = (byte) c[i];
+			this.writeInt(str.length());
+			ensureCapacity(this.idx + str.length() * 2);
+			for(int i = 0; i < str.length(); i++) {
+				final char ch = str.charAt(i);
+				// big-endian
+				this.bytes[this.idx + i * 2] = (byte)((ch >> 8) & 0xff);
+				this.bytes[this.idx + i * 2 + 1] = (byte)(ch & 0xff);
 			}
+			this.idx += str.length() * 2;
```

Comment:
I know the Encoder/ByteBuffer approach made the tests slower, but I'm curious how this loop compares to a more canonical solution like

```java
			byte[] encodedBytes = str.getBytes(StandardCharsets.UTF_8);
			this.writeInt(encodedBytes.length);
			this.ensureCapacity(this.idx + encodedBytes.length);
			System.arraycopy(encodedBytes, 0, this.bytes, this.idx, encodedBytes.length);
			this.idx += encodedBytes.length;
```

Some part of me is still hopeful we can use UTF-8, since it is backward-compatible with TLC's current ASCII encoding and can be more compact on disk.

Reply:
(A _very_ preliminary benchmark I did on my laptop suggests that using `str.getBytes()` over a hand-rolled loop has very little impact on TLC's overall performance---but that result is far from definitive.)

Reply:
Agreed - that would be interesting to see. The optimized code I saw when I looked into the source of `getBytes` suggests the basically the only significant factor might be memory pressure, and it is quite possible that the JVM can optimize such a short-lived local object.

I think I found the benchmarks that @lemmy keeps mentioning: https://github.com/tlaplus/tlaplus/tree/master/tlatools/org.lamport.tlatools/test-benchmark. It looks like they test specifically the subject of this conversation - the disk queue - so it might be relatively straightforward to set up.

I should do some PhD work for a moment, but I can try that on my workstation (or on any other test machines we may or may not have) soon.

Reply:
There are also some end-to-end benchmarks under [general/performance/](https://github.com/tlaplus/tlaplus/tree/master/general/performance). Here is the script I use to compare two Git revisions (must be placed in and run from the root of the tlaplus repo):
https://gist.github.com/Calvin-L/227185db3f7f7131259f195978eccb61

Reply:
I used your script to benchmark the original ASCII encoding, UTF-8 with getBytes, and UTF-16.
I set it to run the full 600s per operation and left it for a while (though it was running on my personal machine so maybe not the cleanest benchmark).

Confusingly, using `getBytes` is noticeably more efficient than the UTF-16 solution. Both are worse than the ASCII version by some factor.

Here are the comparisons results.

before = ASCII, after = UTF-16:
```
Spec                 | W   | States/s (old)  | States/s (new)  | Change    
--------------------------------------------------------------------------------
Grid5k               |   2 |         2256543 |         2446747 |       1.08x FASTER
Ghostferry           |   2 |          130599 |          135148 |       1.03x FASTER
PaxosMadeSimple      |   8 |          119442 |          122801 |       1.03x FASTER
SwarmKit             |   4 |           15901 |           16108 |       1.01x FASTER
EWD840               |   4 |         3451637 |         3444705 |       1.00x SLOWER
MongoRepl            |   2 |          162027 |          160870 |       0.99x SLOWER
PaxosCommit          |   4 |          189351 |          187193 |       0.99x SLOWER
SwarmKit             |   2 |            9648 |            9517 |       0.99x SLOWER
PaxosMadeSimple      |   2 |           52861 |           49065 |       0.93x SLOWER
SwarmKit             |   8 |           26582 |           24314 |       0.91x SLOWER
Grid5k               |   4 |         3269984 |         2957318 |       0.90x SLOWER
EWD840               |   2 |         2204050 |         1965863 |       0.89x SLOWER
PaxosMadeSimple      |   4 |           89819 |           79475 |       0.88x SLOWER
PaxosCommit          |   2 |          151036 |          130710 |       0.87x SLOWER
MongoRepl            |   4 |          261426 |          221925 |       0.85x SLOWER
Bloemen              |   4 |          143643 |          115391 |       0.80x SLOWER
Bloemen              |   2 |          145588 |          115771 |       0.80x SLOWER
Ghostferry           |   8 |          231712 |          169007 |       0.73x SLOWER
MongoRepl            |   8 |          361961 |          263078 |       0.73x SLOWER
Ghostferry           |   4 |          217807 |          153329 |       0.70x SLOWER
Bookkeeper           |   4 |           67047 |           45208 |       0.67x SLOWER
Bookkeeper           |   2 |           58110 |           36756 |       0.63x SLOWER
Bloemen              |   8 |          184246 |          108148 |       0.59x SLOWER
PaxosCommit          |   8 |          306250 |          176044 |       0.57x SLOWER
Bookkeeper           |   8 |           77078 |           40276 |       0.52x SLOWER
```

before = ASCII, after = UTF-8 w/ getBytes
```
Spec                 | W   | States/s (old)  | States/s (new)  | Change    
--------------------------------------------------------------------------------
PaxosMadeSimple      |   8 |          119442 |          131574 |       1.10x FASTER
Ghostferry           |   2 |          130599 |          141453 |       1.08x FASTER
MongoRepl            |   2 |          162027 |          166811 |       1.03x FASTER
Grid5k               |   2 |         2256543 |         2319965 |       1.03x FASTER
SwarmKit             |   4 |           15901 |           16204 |       1.02x FASTER
EWD840               |   4 |         3451637 |         3511398 |       1.02x FASTER
SwarmKit             |   8 |           26582 |           26191 |       0.99x SLOWER
PaxosCommit          |   4 |          189351 |          183080 |       0.97x SLOWER
SwarmKit             |   2 |            9648 |            9169 |       0.95x SLOWER
PaxosMadeSimple      |   2 |           52861 |           49874 |       0.94x SLOWER
Bookkeeper           |   2 |           58110 |           54799 |       0.94x SLOWER
EWD840               |   2 |         2204050 |         2046023 |       0.93x SLOWER
PaxosMadeSimple      |   4 |           89819 |           82986 |       0.92x SLOWER
Grid5k               |   4 |         3269984 |         2985655 |       0.91x SLOWER
PaxosCommit          |   2 |          151036 |          137767 |       0.91x SLOWER
Ghostferry           |   8 |          231712 |          199879 |       0.86x SLOWER
MongoRepl            |   4 |          261426 |          216561 |       0.83x SLOWER
PaxosCommit          |   8 |          306250 |          250167 |       0.82x SLOWER
Bloemen              |   4 |          143643 |          112859 |       0.79x SLOWER
MongoRepl            |   8 |          361961 |          273302 |       0.76x SLOWER
Bookkeeper           |   8 |           77078 |           55919 |       0.73x SLOWER
Bookkeeper           |   4 |           67047 |           47630 |       0.71x SLOWER
Bloemen              |   2 |          145588 |          101009 |       0.69x SLOWER
Ghostferry           |   4 |          217807 |          146945 |       0.67x SLOWER
Bloemen              |   8 |          184246 |           94671 |       0.51x SLOWER
```

before = UTF-16, after = UTF-8 w/ getBytes
```
Spec                 | W   | States/s (old)  | States/s (new)  | Change    
--------------------------------------------------------------------------------
Bookkeeper           |   2 |           36756 |           54799 |       1.49x FASTER
PaxosCommit          |   8 |          176044 |          250167 |       1.42x FASTER
Bookkeeper           |   8 |           40276 |           55919 |       1.39x FASTER
Ghostferry           |   8 |          169007 |          199879 |       1.18x FASTER
SwarmKit             |   8 |           24314 |           26191 |       1.08x FASTER
PaxosMadeSimple      |   8 |          122801 |          131574 |       1.07x FASTER
PaxosCommit          |   2 |          130710 |          137767 |       1.05x FASTER
Bookkeeper           |   4 |           45208 |           47630 |       1.05x FASTER
Ghostferry           |   2 |          135148 |          141453 |       1.05x FASTER
PaxosMadeSimple      |   4 |           79475 |           82986 |       1.04x FASTER
EWD840               |   2 |         1965863 |         2046023 |       1.04x FASTER
MongoRepl            |   8 |          263078 |          273302 |       1.04x FASTER
MongoRepl            |   2 |          160870 |          166811 |       1.04x FASTER
EWD840               |   4 |         3444705 |         3511398 |       1.02x FASTER
PaxosMadeSimple      |   2 |           49065 |           49874 |       1.02x FASTER
Grid5k               |   4 |         2957318 |         2985655 |       1.01x FASTER
SwarmKit             |   4 |           16108 |           16204 |       1.01x FASTER
Bloemen              |   4 |          115391 |          112859 |       0.98x SLOWER
PaxosCommit          |   4 |          187193 |          183080 |       0.98x SLOWER
MongoRepl            |   4 |          221925 |          216561 |       0.98x SLOWER
SwarmKit             |   2 |            9517 |            9169 |       0.96x SLOWER
Ghostferry           |   4 |          153329 |          146945 |       0.96x SLOWER
Grid5k               |   2 |         2446747 |         2319965 |       0.95x SLOWER
Bloemen              |   8 |          108148 |           94671 |       0.88x SLOWER
Bloemen              |   2 |          115771 |          101009 |       0.87x SLOWER
```

I feel there's some more to read into which benchmarks are faster vs slower. It took over 24hrs to get the data though, so I'll put this here and come back to it.

Reply:
Here's a graph of the above:
![image](https://github.com/user-attachments/assets/f67d721c-1753-442a-92be-ef92cc84c59f)

I'm realizing now that error bars would be ideal (if I re-ran this 2x times, for 60hrs runtime total which is a lot of time to not really use my workstation for much), but it seems like the distinction between the 3 versions is unclear. I can imagine it's possible that the error bars are larger than the variation I'm seeing.

There may be some true performance regression, at least on a microbenchmark level, compared to the original (see this discussion, anecdotally https://stackoverflow.com/questions/12239993/why-is-the-native-string-getbytes-method-slower-than-the-custom-implemented-getb). That said, the linked discussion is actually wrong: getBytes has hardcoded special cases for UTF-8 and UTF-16 that have none of the drawbacks mentioned (at least, in my copy of the JDK's source code). I was concerned about the temporary array allocation causing problems, but perhaps the GC / JIT  (maybe escape analysis?) finds a way to make that irrelevant. That certainly seems to be the case, based on the resulting performance of TLC.

So, it seems inconclusive but maybe ok?
I'll commit the getBytes version for now, and if we want further study, then we can do that.